File: eco-core/core-plugin/src/main/java/com/willfp/ecoarmor/sets/ArmorSet.java
package com.willfp.ecoarmor.sets;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.config.interfaces.Config;
import com.willfp.eco.core.display.Display;
import com.willfp.eco.core.items.CustomItem;
import com.willfp.eco.core.items.Items;
import com.willfp.eco.core.items.builder.ItemBuilder;
import com.willfp.eco.core.items.builder.ItemStackBuilder;
import com.willfp.eco.core.recipe.Recipes;
import com.willfp.ecoarmor.sets.util.ArmorUtils;
import com.willfp.ecoarmor.upgrades.Tier;
import com.willfp.ecoarmor.upgrades.Tiers;
import com.willfp.libreforge.Holder;
import com.willfp.libreforge.conditions.Conditions;
import com.willfp.libreforge.conditions.ConfiguredCondition;
import com.willfp.libreforge.effects.ConfiguredEffect;
import com.willfp.libreforge.effects.Effects;
import lombok.AccessLevel;
import lombok.Getter;
import org.bukkit.Bukkit;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.persistence.PersistentDataType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

public class ArmorSet {
    /**
     * Instance of EcoArmor.
     */
    @Getter(AccessLevel.PRIVATE)
    private final EcoPlugin plugin;

    /**
     * The config of the set.
     */
    @Getter(AccessLevel.PRIVATE)
    private final Config config;

    /**
     * The name of the set.
     */
    @Getter
    private final String id;

    /**
     * The advanced holder.
     */
    @Getter
    private final Holder advancedHoler;

    /**
     * The regular holder.
     */
    @Getter
    private final Holder regularHolder;

    /**
     * Items in set.
     */
    private final Map<ArmorSlot, ItemStack> items = new HashMap<>();

    /**
     * Items in advanced set.
     */
    private final Map<ArmorSlot, ItemStack> advancedItems = new HashMap<>();

    /**
     * Advancement shard item.
     */
    @Getter
    private final ItemStack advancementShardItem;

    /**
     * Create a new Armor Set.
     *
     * @param config The set's config.
     * @param plugin Instance of EcoArmor.
     */
    public ArmorSet(@NotNull final Config config,
                    @NotNull final EcoPlugin plugin) {
        this.config = config;
        this.plugin = plugin;
        this.id = config.getString("id");

        Set<ConfiguredCondition> conditions = new HashSet<>();
        for (Config cfg : this.getConfig().getSubsections("conditions")) {
            ConfiguredCondition conf = Conditions.compile(cfg, "Armor Set " + this.id);
            if (conf != null) {
                conditions.add(conf);
            }
        }

        Set<ConfiguredEffect> effects = new HashSet<>();
        for (Config cfg : this.getConfig().getSubsections("effects")) {
            ConfiguredEffect conf = Effects.compile(cfg, "Armor Set " + this.id);
            if (conf != null) {
                effects.add(conf);
            }
        }

        Set<ConfiguredEffect> advancedEffects = new HashSet<>();
        for (Config cfg : this.getConfig().getSubsections("advancedEffects")) {
            ConfiguredEffect conf = Effects.compile(cfg, "Armor Set " + this.id + " (Advanced)");
            if (conf != null) {
                advancedEffects.add(conf);
            }
        }

        this.regularHolder = new RegularHolder(conditions, effects);
        this.advancedHoler = new AdvancedHolder(conditions, advancedEffects);

        ArmorSets.addNewSet(this);

        for (ArmorSlot slot : ArmorSlot.values()) {
            ItemStack item = construct(slot, this.getConfig().getSubsection(slot.name().toLowerCase()), false);
            items.put(slot, item);
            constructRecipe(slot, this.getConfig().getSubsection(slot.name().toLowerCase()), item);

            ItemStack advancedItem = construct(slot, this.getConfig().getSubsection(slot.name().toLowerCase()), true);
            advancedItems.put(slot, advancedItem);
        }

        this.advancementShardItem = constructShard();
    }

    private ItemStack constructShard() {
        List<String> shardLore = this.getConfig().getStrings("advancementShardLore");
        shardLore.replaceAll(s -> Display.PREFIX + s);

        ItemStack shard = new ItemStackBuilder(Items.lookup(this.getPlugin().getConfigYml().getString("advancement-shard-material").toLowerCase()).getItem())
                .setDisplayName(this.getConfig().getString("advancementShardName"))
                .addEnchantment(Enchantment.DURABILITY, 3)
                .addItemFlag(ItemFlag.HIDE_ENCHANTS)
                .addLoreLines(shardLore)
                .writeMetaKey(this.getPlugin().getNamespacedKeyFactory().create("advancement-shard"), PersistentDataType.STRING, id)
                .build();

        if (this.getConfig().getBool("shardCraftable")) {
            Recipes.createAndRegisterRecipe(this.getPlugin(),
                    this.getId() + "_shard",
                    shard,
                    this.getConfig().getStrings("shardRecipe"));
        }

        new CustomItem(
                this.getPlugin().getNamespacedKeyFactory().create("shard_" + id.toLowerCase()),
                test -> this.equals(ArmorUtils.getShardSet(test)),
                shard
        ).register();

        return shard;
    }

    private ItemStack construct(@NotNull final ArmorSlot slot,
                                @NotNull final Config slotConfig,
                                final boolean advanced) {
        ItemBuilder builder = new ItemStackBuilder(Items.lookup(slotConfig.getString("item")).getItem())
                .setDisplayName(advanced ? slotConfig.getString("advancedName") : slotConfig.getString("name"))
                .addLoreLines(slotConfig.getStrings("lore").stream().map(s -> Display.PREFIX + s).collect(Collectors.toList()))
                .addLoreLines(() -> {
                    if (advanced) {
                        return this.getConfig().getStrings("advancedLore").stream().map(s -> Display.PREFIX + s).collect(Collectors.toList());
                    } else {
                        return null;
                    }
                })
                .setDisplayName(() -> advanced ? slotConfig.getString("advancedName") : slotConfig.getString("name"));

        builder.writeMetaKey(
                this.getPlugin().getNamespacedKeyFactory().create("set"),
                PersistentDataType.STRING,
                id
        ).writeMetaKey(
                this.getPlugin().getNamespacedKeyFactory().create("effective-durability"),
                PersistentDataType.INTEGER,
                slotConfig.getInt("effectiveDurability")
        );

        ItemStack itemStack = builder.build();

        ArmorUtils.setAdvanced(itemStack, advanced);
        Tier defaultTier = Tiers.getByID(slotConfig.getString("defaultTier"));
        if (defaultTier == null) {
            Bukkit.getLogger().warning("Default tier specified in " + this.id + " " + slot.name().toLowerCase() + " is invalid! Defaulting to 'default'");
            ArmorUtils.setTier(itemStack, Tiers.getDefaultTier());
        } else {
            ArmorUtils.setTier(itemStack, defaultTier);
        }

        if (advanced) {
            new CustomItem(this.getPlugin().getNamespacedKeyFactory().create("set_" + id.toLowerCase() + "_" + slot.name().toLowerCase() + "_advanced"), test -> {
                if (ArmorSlot.getSlot(test) != ArmorSlot.getSlot(itemStack)) {
                    return false;
                }
                if (!ArmorUtils.isAdvanced(itemStack)) {
                    return false;
                }
                if (ArmorUtils.getSetOnItem(test) == null) {
                    return false;
                }

                return Objects.equals(this, ArmorUtils.getSetOnItem(test));
            }, itemStack).register();
        } else {
            new CustomItem(this.getPlugin().getNamespacedKeyFactory().create("set_" + id.toLowerCase() + "_" + slot.name().toLowerCase()), test -> {
                if (ArmorSlot.getSlot(test) != ArmorSlot.getSlot(itemStack)) {
                    return false;
                }
                if (ArmorUtils.isAdvanced(itemStack)) {
                    return false;
                }
                if (ArmorUtils.getSetOnItem(test) == null) {
                    return false;
                }

                return Objects.equals(this, ArmorUtils.getSetOnItem(test));
            }, itemStack).register();
        }

        return itemStack;
    }

    private void constructRecipe(@NotNull final ArmorSlot slot,
                                 @NotNull final Config slotConfig,
                                 @NotNull final ItemStack out) {
        if (slotConfig.getBool("craftable")) {
            ItemStack formattedOut = out.clone();
            ItemMeta meta = formattedOut.getItemMeta();
            assert meta != null;
            assert meta.getLore() != null;

            List<String> lore = new ArrayList<>();

            for (String s : meta.getLore()) {
                s = s.replace("%tier%", Tiers.getDefaultTier().getDisplayName());
                lore.add(s);
            }

            meta.setLore(lore);
            formattedOut.setItemMeta(meta);

            Recipes.createAndRegisterRecipe(
                    this.getPlugin(),
                    this.getId() + "_" + slot.name().toLowerCase(),
                    formattedOut,
                    slotConfig.getStrings("recipe")
            );
        }
    }

    /**
     * Get item stack from slot.
     *
     * @param slot The slot.
     * @return The item.
     */
    public ItemStack getItemStack(@NotNull final ArmorSlot slot) {
        return items.get(slot);
    }

    /**
     * Get item stack from slot.
     *
     * @param slot The slot.
     * @return The item.
     */
    public ItemStack getAdvancedItemStack(@NotNull final ArmorSlot slot) {
        return advancedItems.get(slot);
    }

    /**
     * Get default tier for slot.
     *
     * @param slot The slot.
     * @return The tier.
     */
    public Tier getDefaultTier(@Nullable final ArmorSlot slot) {
        if (slot == null) return Tiers.getDefaultTier();
        Tier tier = Tiers.getByID(this.config.getSubsection(slot.name().toLowerCase()).getString("defaultTier"));
        return tier != null ? tier : Tiers.getDefaultTier();
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }

        if (!(o instanceof ArmorSet set)) {
            return false;
        }

        return this.id.equals(set.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.id);
    }

    @Override
    public String toString() {
        return "ArmorSet{"
                + this.id
                + "}";
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecoarmor/sets/ArmorSet.kt
package com.willfp.ecoarmor.sets

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.config.interfaces.Config
import com.willfp.eco.core.display.Display
import com.willfp.eco.core.items.CustomItem
import com.willfp.eco.core.items.Items
import com.willfp.eco.core.items.builder.ItemBuilder
import com.willfp.eco.core.items.builder.ItemStackBuilder
import com.willfp.eco.core.recipe.Recipes
import com.willfp.eco.core.registry.Registrable
import com.willfp.ecoarmor.sets.ArmorSlot.Companion.getSlot
import com.willfp.ecoarmor.sets.ArmorUtils.getSetOnItem
import com.willfp.ecoarmor.sets.ArmorUtils.getShardSet
import com.willfp.ecoarmor.sets.ArmorUtils.isAdvanced
import com.willfp.ecoarmor.sets.ArmorUtils.setAdvanced
import com.willfp.ecoarmor.sets.ArmorUtils.setTier
import com.willfp.ecoarmor.upgrades.Tier
import com.willfp.ecoarmor.upgrades.Tiers
import com.willfp.ecoarmor.util.PlayableSound
import com.willfp.libreforge.Holder
import com.willfp.libreforge.ItemProvidedHolder
import com.willfp.libreforge.ProvidedHolder
import com.willfp.libreforge.SimpleHolder
import com.willfp.libreforge.ViolationContext
import com.willfp.libreforge.conditions.Conditions
import com.willfp.libreforge.effects.Effects
import com.willfp.libreforge.notNullMapOf
import com.willfp.libreforge.notNullMutableMapOf
import org.bukkit.Bukkit
import org.bukkit.Sound
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType
import java.util.Locale
import java.util.Objects
import java.util.stream.Collectors

class ArmorSet(
    val id: String,
    val config: Config,
    private val plugin: EcoPlugin
) : Registrable {
    /** The advanced holder. */
    val advancedHolder: Holder

    /** The regular holder. */
    val regularHolder: Holder

    /** Items in set. */
    private val items = notNullMutableMapOf<ArmorSlot, ItemStack>()

    /** Holders in set. */
    private val slotHolders = notNullMutableMapOf<ArmorSlot, Holder>()

    /** Items in advanced set. */
    private val advancedItems = notNullMutableMapOf<ArmorSlot, ItemStack>()

    /** Holders in advanced set. */
    private val advancedSlotHolders = notNullMutableMapOf<ArmorSlot, Holder>()

    /** Advancement shard item. */
    val advancementShardItem: ItemStack

    /*
    * Equip Sound
     */
    val equipSound = if (config.getBool("sounds.equip.enabled")) {
        PlayableSound(
            Sound.valueOf(config.getString("sounds.equip.sound").uppercase()),
            config.getDouble("sounds.equip.volume"),
            config.getDouble("sounds.equip.pitch")
        )
    } else null

    /*
    * Advanced equip Sound
     */
    val advancedEquipSound = if (config.getBool("sounds.advancedEquip.enabled")) {
        PlayableSound(
            Sound.valueOf(config.getString("sounds.advancedEquip.sound").uppercase()),
            config.getDouble("sounds.advancedEquip.volume"),
            config.getDouble("sounds.advancedEquip.pitch")
        )
    } else null

    /*
    * Unequip Sound
     */
    val unequipSound = if (config.getBool("sounds.unequip.enabled")) {
        PlayableSound(
            Sound.valueOf(config.getString("sounds.unequip.sound").uppercase()),
            config.getDouble("sounds.unequip.volume"),
            config.getDouble("sounds.unequip.pitch")
        )
    } else null

    /** Create a new Armor Set. */
    init {
        val conditions = Conditions.compile(
            config.getSubsections("conditions"),
            ViolationContext(plugin, "Armor Set $id")
        )

        val effects = Effects.compile(
            config.getSubsections("effects"),
            ViolationContext(plugin, "Armor Set $id")
        )

        val advancedEffects = Effects.compile(
            config.getSubsections("advancedEffects"),
            ViolationContext(plugin, "Armor Set $id (advanced)")
        )

        regularHolder = SimpleHolder(plugin.namespacedKeyFactory.create(id), effects, conditions)
        advancedHolder = SimpleHolder(plugin.namespacedKeyFactory.create("${id}_advanced"), advancedEffects, conditions)

        for (slot in ArmorSlot.values()) {
            val slotConfig = config.getSubsection(slot.name.lowercase(Locale.getDefault()))
            val item = construct(slot, slotConfig, false)
            items[slot] = item
            constructRecipe(slot, slotConfig, item)
            val advancedItem = construct(slot, slotConfig, true)
            advancedItems[slot] = advancedItem

            slotHolders[slot] = SimpleHolder(
                plugin.createNamespacedKey("${id}_${slot.name.lowercase()}"),
                Effects.compile(
                    slotConfig.getSubsections("effects"),
                    ViolationContext(plugin, "Armor Set $id)")
                        .with(slot.name.lowercase())
                ),
                Conditions.compile(
                    slotConfig.getSubsections("conditions"),
                    ViolationContext(plugin, "Armor Set $id)")
                        .with(slot.name.lowercase())
                )
            )

            advancedSlotHolders[slot] = SimpleHolder(
                plugin.createNamespacedKey("${id}_${slot.name.lowercase()}_advanced"),
                Effects.compile(
                    slotConfig.getSubsections("advancedEffects"),
                    ViolationContext(plugin, "Armor Set $id (advanced)")
                        .with(slot.name.lowercase())
                ),
                Conditions.compile(
                    slotConfig.getSubsections("conditions"),
                    ViolationContext(plugin, "Armor Set $id (advanced)")
                        .with(slot.name.lowercase())
                )
            )
        }

        advancementShardItem = constructShard()
    }

    private fun constructShard(): ItemStack {
        val shardLore = config.getStrings("shard.lore")
        shardLore.replaceAll { Display.PREFIX + it }
        val shard = ItemStackBuilder(
            Items.lookup(config.getString("shard.item"))
        )
            .setDisplayName(config.getFormattedString("shard.name"))
            .addLoreLines(shardLore)
            .writeMetaKey(plugin.namespacedKeyFactory.create("advancement-shard"), PersistentDataType.STRING, id)
            .build()
        if (config.getBool("shard.craftable")) {
            Recipes.createAndRegisterRecipe(
                plugin,
                id + "_shard",
                shard,
                config.getStrings("shard.recipe")
            )
        }
        CustomItem(
            plugin.namespacedKeyFactory.create("shard_" + id.lowercase(Locale.getDefault())),
            { test: ItemStack? -> this == getShardSet(test!!) },
            shard
        ).register()
        return shard
    }

    private fun construct(
        slot: ArmorSlot,
        slotConfig: Config,
        advanced: Boolean
    ): ItemStack {
        val builder: ItemBuilder = ItemStackBuilder(Items.lookup(slotConfig.getString("item")).item).apply {
            setDisplayName(
                if (advanced) slotConfig.getFormattedString("advancedName") else slotConfig.getFormattedString(
                    "name"
                )
            )
            val defaultLore = slotConfig.getFormattedStrings("lore").stream().map { s: String -> Display.PREFIX + s }
                .collect(Collectors.toList())
            val advancedLore = config.getFormattedStrings("advancedLore").stream()
                .map { s: String -> Display.PREFIX + s }
                .collect(Collectors.toList())

            if (advanced) {
                if (!plugin.configYml.getBool("advanced-lore-only")) {
                    addLoreLines(defaultLore)
                }
                addLoreLines(advancedLore)
            } else {
                addLoreLines(defaultLore)
            }

            setDisplayName {
                if (advanced) slotConfig.getFormattedString("advancedName") else slotConfig.getFormattedString(
                    "name"
                )
            }
        }
        builder.writeMetaKey(
            plugin.namespacedKeyFactory.create("set"),
            PersistentDataType.STRING,
            id
        ).writeMetaKey(
            plugin.namespacedKeyFactory.create("effective-durability"),
            PersistentDataType.INTEGER,
            slotConfig.getInt("effectiveDurability")
        )
        val itemStack = builder.build()
        setAdvanced(itemStack, advanced)
        val defaultTier = Tiers.getByID(slotConfig.getString("defaultTier"))
        if (defaultTier == null) {
            Bukkit.getLogger()
                .warning("Default tier specified in " + id + " " + slot.name.lowercase(Locale.getDefault()) + " is invalid! Defaulting to 'default'")
            setTier(itemStack, Tiers.defaultTier)
        } else {
            setTier(itemStack, defaultTier)
        }
        if (advanced) {
            CustomItem(
                plugin.namespacedKeyFactory.create(
                    "set_" + id.lowercase(Locale.getDefault()) + "_" + slot.name.lowercase(
                        Locale.getDefault()
                    ) + "_advanced"
                ), { test ->
                    if (getSlot(test) !== getSlot(itemStack)) {
                        return@CustomItem false
                    }
                    if (!isAdvanced(test)) {
                        return@CustomItem false
                    }
                    if (getSetOnItem(test) == null) {
                        return@CustomItem false
                    }
                    this == getSetOnItem(test)
                }, itemStack
            ).register()
        } else {
            CustomItem(
                plugin.namespacedKeyFactory.create(
                    "set_" + id.lowercase(Locale.getDefault()) + "_" + slot.name.lowercase(
                        Locale.getDefault()
                    )
                ),
                { test ->
                    if (getSlot(test) !== getSlot(itemStack)) {
                        return@CustomItem false
                    }
                    if (isAdvanced(test)) {
                        return@CustomItem false
                    }
                    if (getSetOnItem(test) == null) {
                        return@CustomItem false
                    }
                    this == getSetOnItem(test)
                }, itemStack
            ).register()
        }
        return itemStack
    }

    @Suppress("DEPRECATION")
    private fun constructRecipe(
        slot: ArmorSlot,
        slotConfig: Config,
        out: ItemStack
    ) {
        if (slotConfig.getBool("craftable")) {
            val formattedOut = out.clone()
            val meta = formattedOut.itemMeta ?: return
            val metaLore = meta.lore ?: emptyList()
            val lore = metaLore.map { it.replace("%tier%", Tiers.defaultTier.displayName) }
            meta.lore = lore
            formattedOut.itemMeta = meta
            plugin.scheduler.run {
                Recipes.createAndRegisterRecipe(
                    plugin,
                    id + "_" + slot.name.lowercase(Locale.getDefault()),
                    formattedOut,
                    slotConfig.getStrings("recipe")
                )
            }
        }
    }

    /**
     * Get item stack from slot.
     *
     * @param slot The slot.
     * @return The item.
     */
    fun getItemStack(slot: ArmorSlot): ItemStack {
        return items[slot]
    }

    /**
     * Get item stack from slot.
     *
     * @param slot The slot.
     * @return The item.
     */
    fun getAdvancedItemStack(slot: ArmorSlot): ItemStack {
        return advancedItems[slot]
    }

    /**
     * Get default tier for slot.
     *
     * @param slot The slot.
     * @return The tier.
     */
    fun getDefaultTier(slot: ArmorSlot?): Tier {
        slot ?: return Tiers.defaultTier
        val tier = Tiers.getByID(config.getSubsection(slot.name.lowercase()).getString("defaultTier"))
        return tier ?: Tiers.defaultTier
    }

    fun getSpecificHolder(itemStack: ItemStack): ItemProvidedHolder? {
        val slot = getSlot(itemStack) ?: return null
        val advanced = isAdvanced(itemStack)

        return if (advanced) {
            ItemProvidedHolder(advancedSlotHolders[slot], itemStack)
        } else {
            ItemProvidedHolder(slotHolders[slot], itemStack)
        }
    }

    override fun getID(): String {
        return this.id
    }

    override fun equals(other: Any?): Boolean {
        if (other !is ArmorSet) {
            return false
        }

        return id == other.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }

    override fun toString(): String {
        return ("ArmorSet{"
                + id
                + "}")
    }
}
